### PR TITLE
int24/uint24 support

### PR DESCRIPTION
**Pull request by sunng87**
Corrected re-implementation of #3

### CODE REVIEW

File: src/gloss/data/primitives.clj [hunk 1]
```diff
@@ -71,6 +71,51 @@
...
+    (bytes->umedium bytes :le)))
+
+(defn put-medium [buffer value]
+  (let [bytes (medium->bytes value :be)]
+    (.put ^ByteBuffer buffer ^bytes bytes)))
+
+(defn put-medium-le [buffer value]
+  (let [bytes (medium->bytes value :le)]
+    (.put ^ByteBuffer buffer ^bytes bytes)))
+
```
Comment (KingMob) [inline src/gloss/data/primitives.clj:118]: Hmm, won't the default versions fail if the ByteBuffer's ordering is LE? The built-in versions of things like `putInt` and `getLong` use the `.order()` of the ByteBuffer to interpret the next _X_ bytes, but `get-unsigned-medium` and `put-medium` assume they're big-endian, iiuc.

File: test/gloss/test/core.clj [hunk 2]
```diff
@@ -314,3 +314,8 @@
         dec)
       (string :utf-8 :suffix ","))
     "abc"))
+
+(deftest test-int24s
+  (test-roundtrip
+   [:int24 :int24-le :int24-be :uint24 :uint24-le :uint24-be]
+   [32779 32779 32779 32779 32779 32779]))
```
Comment (KingMob) [inline test/gloss/test/core.clj:321]: Can you expand this a bit? Can we test negative numbers with the signed versions, and both BE and LE-ordered ByteBuffers?

**Comment by KingMob** [comment]
Thanks for doing this!

File: src/gloss/data/primitives.clj (hunk 1 above)
Comment (sunng87) [reply src/gloss/data/primitives.clj]: I just refactored to use native byte order for default codec.

File: test/gloss/test/core.clj (hunk 2 above)
Comment (sunng87) [reply test/gloss/test/core.clj]: Just add bytebuffer output check.

File: src/gloss/data/primitives.clj [hunk 3]
```diff
@@ -148,6 +198,13 @@
    :uint16-le (primitive-codec .getShort .putShort 2 short->ushort short ushort->short :le)
    :uint16-be (primitive-codec .getShort .putShort 2 short->ushort short ushort->short :be)
 
+   :uint24 (primitive-codec (partial get-unsigned-medium (native-byte-order))
+                            (partial put-medium (native-byte-order)) 3 identity int identity)
+   :uint24-le (primitive-codec (partial get-unsigned-medium :le)
+                               (partial put-medium :le) 3 identity int identity :le)
```
Comment (sunng87) [inline src/gloss/data/primitives.clj:204]: Not sure if the last `:le` here still needed because I'm dealing with the byteorder manually

File: src/gloss/data/primitives.clj [hunk 4]
```diff
@@ -71,6 +71,47 @@
    :be `ByteOrder/BIG_ENDIAN
    :ne `(ByteOrder/nativeOrder)})
 
+(defn- native-byte-order []
+  (if (= (ByteOrder/nativeOrder) ByteOrder/LITTLE_ENDIAN)
+    :le
+    :be))
```
Comment (KingMob) [inline src/gloss/data/primitives.clj:77]: This can be a `def`, not a `defn`, since `ByteOrder/NATIVE_ORDER` is `final`.

File: src/gloss/data/primitives.clj (hunk 3 above)
Comment (KingMob) [reply src/gloss/data/primitives.clj:204]: Yeah, it might not be. For the built-in supported types, the final `bo` parameter is used to force the buf to a certain order, write out the value in the current order, and then flip the order back, to force writing LE values in a BE buffer (or vice versa).

But you're manually writing out the bytes in sequential order, and `medium->bytes/bytes->umedium` take care of the order before writing, flipping the order should have no effect. I think you can leave the `:le` and `:be` out, but add an explanatory comment above, saying _why_ they're not necessary in this case.

File: test/gloss/test/core.clj [hunk 5]
```diff
@@ -314,3 +314,26 @@
...
+     (repeat 3 -1924)))
+  (testing "checking results against netty codec output"
+    (= [-20 -5 -1] (.array (first (encode [:int24-le] [-1044]))))
+    (= [-1 -5 -20] (.array (first (encode [:int24-be] [-1044]))))
+    (= [11 -128 0] (.array (first (encode [:int24-le] [32779]))))
+    (= [0 -128 11] (.array (first (encode [:int24-be] [32779]))))
+    (= [11 -128 0] (.array (first (encode [:uint24-le] [32779]))))
+    (= [0 -128 11] (.array (first (encode [:uint24-be] [32779]))))
+    (= [-1 -1 63] (.array (first (encode [:uint24-le] [4194303]))))
+    (= [63 -1 -1] (.array (first (encode [:uint24-be] [4194303]))))))
```
Comment (KingMob) [inline test/gloss/test/core.clj:339]: Can you add some tests of the maximum and minimum uint24 and int24 values? Always gotta check the edges...

**Review by KingMob — CHANGES_REQUESTED** [review]
Almost done, I think. It's looking good.

File: src/gloss/data/primitives.clj (hunk 4 above)
Comment (sunng87) [reply src/gloss/data/primitives.clj]: Fixed

**Review by KingMob — APPROVED** [review]
👍🏻

**Comment by KingMob** [comment]
Done! Thanks so much @sunng87

**Comment by KingMob** [comment]
Fixes #2
